### PR TITLE
Set `clusterIP` field for kubevirt-prometheus-metrics service to `None`

### PR DESCRIPTION
**Pull request by VirrageS**
**What this PR does / why we need it**: `kubevirt-prometheus-metrics` Service must be created with `ClusterIP` set to `None`. If not, we end up creating a LoadBalancer rule with endpoints equal to number of nodes in the cluster. If we have a 1000-node cluster, then we will have a LB rule with 1000 endpoints. For more info see: https://kubernetes.io/docs/concepts/services-networking/service/#headless-services

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: None

**Special notes for your reviewer**: None

**Release note**:

```release-note
Service `kubevirt-prometheus-metrics` now sets `ClusterIP` to `None` to make it a headless service.
```

### CODE REVIEW

**Comment by VirrageS** [comment]
@maiqueb could you take a look?

**Review by maiqueb — CHANGES_REQUESTED** [review]
Code wise, I'm OK.

... but you need to add at least a unit test for this. Something that defines the expected behavior of the `kubevirt-prometheus-metrics` service to be a headless service.

Unfortunately, this pkg doesn't exhibit this good practice, but, if you don't mind (too much...) I would rather have you start it :)

**Comment by VirrageS** [comment]
I've added a test, but not entirely sure what should I check more. Let me know if you have something specific in mind.

**Review by maiqueb — CHANGES_REQUESTED** [review]
You'll also need to update the bazel metadata (you've added a file); run `make deps-update` (and to be on the safe side also `make generate`), and commit those changes.

File: pkg/virt-operator/resource/generate/components/deployments_test.go
Comment (VirrageS) [reply pkg/virt-operator/resource/generate/components/deployments_test.go]: Whoops, I've copied wrong variable. Sorry for that, I've fixed that and tests now pass.

**Comment by VirrageS** [comment]
@maiqueb thanks for catching all of that. I've updated the MR. Let me know if there is something more to be done.

**Review by maiqueb — APPROVED** [review]
Thank you.

**Comment by VirrageS** [comment]
@vladikr could you take a look?

**Comment by maiqueb** [comment]
/remove-lifecycle stale

@xpivarc / @AlonaKaplan could you take a look into this ?

**Comment by VirrageS** [comment]
@sradco @Barakmor1 could you take a look and approve if looks okay?

**Comment by Barakmor1** [comment]
/lgtm
Thank you

**Comment by xpivarc** [comment]
@VirrageS I am less confident in this area but I guess you want to optimize this. Prometheus will only use the endpoints for scrapping the data, is that right? (Meaning it does not need the static IP)
/test ?

**Comment by Barakmor1** [comment]
/hold
@VirrageS just few questions:
> we end up creating a LoadBalancer rule with endpoints equal to number of nodes in the cluster. If we have a 1000-node 
> cluster, then we will have a LB rule with 1000 endpoints. 
1) why would we have 1000 endpoints?
   I think that the number of endpoints will be equal to the number of pods backing the service

2) The LB will make sure we divide the service workload to all the pods, why do we need to remove it?

**Comment by VirrageS** [comment]
> why would we have 1000 endpoints?
> I think that the number of endpoints will be equal to the number of pods backing the service
Yes, and I had in mind that we have 1000 nodes with `virt-handlers` running on them.

> The LB will make sure we divide the service workloads to all the pods, why do we need to remove it?
But why do you need to divide the workload if you don't have any workload? Prometheus need to ask each Pod individually, it doesn't send request to the Service itself. It uses Service to get all IPs of the Pods backing the Service. Otherwise, how the Prometheus would get the metrics from all the Pods? (if the load-balancer would be random it could be weeks until some specific Pod gets the request).

**Comment by Barakmor1** [comment]
> @Barakmor1
> 
> > why would we have 1000 endpoints?
> > I think that the number of endpoints will be equal to the number of pods backing the service
> 
> Yes, and I had in mind that we have 1000 nodes with 1000 `virt-handlers` running on them (each `virt-handler` on each node).
> 
> > The LB will make sure we divide the service workloads to all the pods, why do we need to remove it?
> 
> But why do you need to divide the workload if you don't have any workload? Prometheus need to ask each Pod individually, it doesn't send request to the Service itself. It uses Service to get all IPs of the Pods backing the Service. Otherwise, how the Prometheus would get the metrics from all the Pods? (if the load-balancer would be random it could be weeks until some specific Pod gets the request).

Got it Thank you for the explaination.
last question:
What happens now when we do have 1000 nodes in the cluster how does Prometheus get metrics from all the virt-handler pods?

**Comment by Barakmor1** [comment]
Maybe it would be better using PodMonitors 
https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/getting-started.md#using-podmonitors

**Comment by rmohr** [comment]
> I can explore that in follow-up MR. I would need to do the testing, I wasn't aware of `PodMonitor`s before.

Now with `None`, it should be equally efficient. I am not sure a switch is necessary.

**Pull request merged**
